lua: allow to override a built-in module

Totktonada opened the pull request:
This patchset continues the series (#8173, #8250) of commits about overriding built-in modules (#7774). It finally solves the main task. The changes:

* Use `luaT_setmodule()` to register serializers.
* Use another table instead of `package.loaded` to store built-in modules (and add a loader that looks into this table).
* Set `_G.arg` earlier to make it accessible from override modules.
* Added `gentree` and `justrun` testing helpers (see commits for details).
* Added `override.<module_name>` loader.
* Added an ability to enable/disable the override loader.

Several things are technically out of scope of the issue, but they should be solved together with it:

* Add a mechanism for discarding assignment to `package.loaded` for a module.
* Add a flag to indicate, whether tarantool's initialization is finished.
* Search for `.rocks` next to the main script file.

I'll send them in separate pull requests.

Fixes #7774

igormunkin reviewed (approved):
**[PATCH 1/7] lua: register serializers using luaT_setmodule()**

@Totktonada, thanks for the patch! LGTM as trivial.

igormunkin commented on test/luajit-test-init.lua:
```diff
@@ -1,9 +1,12 @@
+
 -- XXX: lua-Harness test suite uses it's own tap.lua module
 -- that conflicts with the Tarantool's one.
 package.loaded.tap = nil
```
Minor: I believe, that as a result of the corresponding patches removing builtin modules from `package.loaded`, this line can be removed, right?

Totktonada replied:
Technically, I guess, yes. However it would be fragile.

If there is no `require('tap')` call during tarantool's initialization, `package.loaded.tap` is `nil`. We only need `loaders.builtin.tap = nil` to choose an external module at next `require('tap')`.

However if some built-in module will depend on `tap` (and so call `require('tap')`), the built-in `tap` module will be written to `package.loaded.tap` and future require() will return it.

So the old `package.loaded.foo = nil` hack is now `package.loaded.foo = nil` plus `loaders.builtin.foo = nil`.

igormunkin replied:
Alright, let's leave it as is then.

igormunkin commented on src/lua/init.c:
```diff
@@ -839,6 +839,13 @@ tarantool_lua_init(const char *tarantool_bin, int argc, char **argv)
+	 * Create a table for storing loaded built-in modules.
+	 * Similar to _LOADED (package.loaded).
+	 */
+	lua_createtable(L, 0, 128);
```
Minor: Why 128? I would rather use "`sizeof(lua_modules)`" for this.

Totktonada replied:
`lua_modules` is mostly irrelevant here. There are modules written using C. There are box's modules.

I did some calculations and got a number around 90 or 100 for this table. So 128 looks good to avoid rehashing and to don't have reasons to touch this for a long time. I can set it to zero if it is important. I definitelly will not write anything more complex than a constant here.

igormunkin replied:
OK, I get it now, thanks. Anyway, I would like to hear @Buristan opinion here. The most bothering thing here is the raw constant, but I can try to live with it :)

Totktonada replied:
Changed to `lua_newtable(L)`. No magic constants anymore.

igormunkin replied:
Thanks!

Buristan reviewed (commented):
Hi, Alexander!
Thanks for the patch-set!
I'll review it per patch.

**[PATCH 1/7] lua: register serializers using luaT_setmodule()**
LGTM!

igormunkin commented on src/lua/minifio.c:
```diff
@@ -53,6 +53,12 @@ tarantool_lua_minifio_init(struct lua_State *L)
+	 * The loaders.builtin loader is not in effect yet.
+	 * Set the minifio module into package.loaded manually.
+	 */
+	lua_getfield(L, LUA_REGISTRYINDEX, "_LOADED");
```
Minor: why do we need use `_LOADED` here instead of `_PRELOAD`?

Totktonada replied:
I have several answers here.

1. Why should we use `_PRELOAD` if we need only `_LOADED`? If we would use `_PRELOAD` here, the result would be setting the value into `_LOADED` in a bit less direct way.
2. If there would be the builtin loader at this stage, `require()` would find the module and store it in `_LOADED`. We imitate this behavior.
3. `_PRELOAD` is accessible by a user and I prefer to don't touch it in the platform if possible.

igormunkin replied:
1. `package.loaded[module] = nil` is a standard idiom for module "unload"; in general case `package.preload` allows to reload the module if this is needed.
2. Yeah, but see the reason above.
3. Don't get this point. `package.loaded` is accessible too, isn't it?

Totktonada replied:
We agreed that we can't reach an agreement here, so I don't know. I'll leave it as is.

igormunkin commented on test/unit/luaT_tuple_new.c:
```diff
@@ -187,6 +187,10 @@ main()
 	struct lua_State *L = luaL_newstate();
 	luaL_openlibs(L);
 
+	/* luaT_newmodule() assumes that this table exists. */
```
Side note: I want to share here the approach used in [uJIT tests](https://github.com/luavela/luavela/blob/master/tests/impl/uJIT-tests-C/suite/test_common_lua.h#L27-L35), but I don't ask to fix it within this patchset. If this works for you, feel free to file an issue to move all unit tests for this machinery.

Totktonada replied:
Something like this?

```diff
diff --git a/test/unit/luaT_tuple_new.c b/test/unit/luaT_tuple_new.c
index 0b3847347..42643464c 100644
--- a/test/unit/luaT_tuple_new.c
+++ b/test/unit/luaT_tuple_new.c
@@ -14,6 +14,7 @@
 
 #define UNIT_TAP_COMPATIBLE 1
 #include "unit.h"             /* plan, header, footer, is, ok */
+#include "lua_test_utils.h"
 
 /*
  * This test checks all usage cases of luaT_tuple_new():
@@ -184,12 +185,7 @@ main()
 	memory_init();
 	fiber_init(fiber_c_invoke);
 
-	struct lua_State *L = luaL_newstate();
-	luaL_openlibs(L);
-
-	/* luaT_newmodule() assumes that this table exists. */
-	lua_newtable(L);
-	lua_setfield(L, LUA_REGISTRYINDEX, "_TARANTOOL_BUILTIN");
+	struct lua_State *L = luaT_newteststate();
 
 	box_init();
 	tarantool_lua_error_init(L);
diff --git a/test/unit/lua_msgpack.c b/test/unit/lua_msgpack.c
index 106097638..628ad1259 100644
--- a/test/unit/lua_msgpack.c
+++ b/test/unit/lua_msgpack.c
@@ -14,6 +14,7 @@
 
 #define UNIT_TAP_COMPATIBLE 1
 #include "unit.h"
+#include "lua_test_utils.h"
 
 static void
 mpstream_error_mock(void *ctx)
@@ -151,16 +152,11 @@ main(void)
 	plan(2);
 	header();
 
-	struct lua_State *L = luaL_newstate();
-	luaL_openlibs(L);
+	struct lua_State *L = luaT_newteststate();
 	tarantool_L = L;
 	memory_init();
 	fiber_init(fiber_c_invoke);
 
-	/* luaT_newmodule() assumes that this table exists. */
-	lua_newtable(L);
-	lua_setfield(L, LUA_REGISTRYINDEX, "_TARANTOOL_BUILTIN");
-
 	tarantool_lua_serializer_init(L);
 	luaopen_msgpack(L);
 	lua_pop(L, 1);
diff --git a/test/unit/lua_test_utils.h b/test/unit/lua_test_utils.h
new file mode 100644
index 000000000..1ed5dcb36
--- /dev/null
+++ b/test/unit/lua_test_utils.h
@@ -0,0 +1,36 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#pragma once
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lualib.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/**
+ * Incapsulates steps we should do to create a Lua state suitable
+ * to initialize some built-in Lua modules.
+ */
+static inline struct lua_State *
+luaT_newteststate(void)
+{
+	struct lua_State *L = luaL_newstate();
+	luaL_openlibs(L);
+
+	/* luaT_newmodule() assumes that this table exists. */
+	lua_newtable(L);
+	lua_setfield(L, LUA_REGISTRYINDEX, "_TARANTOOL_BUILTIN");
+
+	return L;
+}
+
+#if defined(__cplusplus)
+} /* extern "C" */
+#endif /* defined(__cplusplus) */
diff --git a/test/unit/lua_tweaks.c b/test/unit/lua_tweaks.c
index 56aa2b229..c5aa46253 100644
--- a/test/unit/lua_tweaks.c
+++ b/test/unit/lua_tweaks.c
@@ -15,6 +15,7 @@
 
 #define UNIT_TAP_COMPATIBLE 1
 #include "unit.h"
+#include "lua_test_utils.h"
 
 static struct lua_State *L;
 
@@ -239,13 +240,8 @@ main(void)
 {
 	memory_init();
 	fiber_init(fiber_c_invoke);
-	L = luaL_newstate();
-	luaL_openlibs(L);
-
-	/* luaT_newmodule() assumes that this table exists. */
-	lua_newtable(L);
-	lua_setfield(L, LUA_REGISTRYINDEX, "_TARANTOOL_BUILTIN");
 
+	L = luaT_newteststate();
 	tarantool_lua_tweaks_init(L);
 
 	/*
```

Added.

igormunkin replied:
Precisely, thanks!

igormunkin commented on test/unit/lua_tweaks.c:
```diff
@@ -241,7 +241,30 @@ main(void)
+
 	tarantool_lua_error_init(L);
 	fail_unless(luaT_dostring(
 			L, "tweaks = require('internal.tweaks')") == 0);
```
I just wonder, whether this line doesn't solve the problem described above?

Totktonada replied:
Agreed to add a bit more info into the comment above.

Totktonada replied:
Added:

> In particular, tarantool_lua_init() funtcion is not called in a unit test.

igormunkin replied:
Great, thanks!

igormunkin commented on src/lua/minifio.c:
```diff
@@ -53,6 +53,12 @@ tarantool_lua_minifio_init(struct lua_State *L)
+	lua_getfield(L, LUA_REGISTRYINDEX, "_LOADED");
 	luaT_newmodule(L, "internal.minifio", minifio_methods);
+	lua_setfield(L, -2, "internal.minifio");
 	lua_pop(L, 1);
```
`lua_pop` looks excess here (`lua_setfield` removes the module field from the stack).

Totktonada replied:
It pops `_LOADED`.

igormunkin replied:
PEBKAC. Anyway, comment desired, if you don't mind.

Totktonada replied:
No problem, added.

```diff
diff --git a/src/lua/minifio.c b/src/lua/minifio.c
index 952ad86e2..7795408a8 100644
--- a/src/lua/minifio.c
+++ b/src/lua/minifio.c
@@ -60,5 +60,5 @@ tarantool_lua_minifio_init(struct lua_State *L)
        lua_getfield(L, LUA_REGISTRYINDEX, "_LOADED");
        luaT_newmodule(L, "internal.minifio", minifio_methods);
        lua_setfield(L, -2, "internal.minifio");
-       lua_pop(L, 1);
+       lua_pop(L, 1); /* _LOADED */
 }
```

igormunkin replied:
Thanks!

igormunkin reviewed (changes requested):
**[PATCH 2/7] lua: don't set built-in modules to package.loaded**

@Totktonada, thanks for the patch! Please consider my comments and questions below.

igormunkin reviewed (approved):
**[PATCH 3/7] lua: define _G.arg earlier**

@Totktonada, thanks for the patch! Looks legit, LGTM.

igormunkin reviewed (approved):
**[PATCH 4/7] test: add a generator for a tree of Lua files**

@Totktonada, thanks for the patch! LGTM.

igormunkin reviewed (approved):
**[PATCH 5/7] test: add a simple tarantool runner/output catcher**

@Totktonada, thanks for the patch! LGTM.

Buristan reviewed (commented):
**[PATCH 2/7] lua: don't set built-in modules to package.loaded**
Generally LGTM, just few minor suggestions below.


> If `foo` is already assigned to `package.loaded`, the loaders will not
called at all, so we can't check the file system.

Typo: s/will not called/will not be called/

>   built-in tap module. However now it is not enough to remove it from

Typo: s/However/However,/

Buristan reviewed (commented):
**[PATCH 3/7] lua: define _G.arg earlier**
LGTM, except a single nit regarding the commit message.
> However nothing prevents us from providing the `arg` global variable at

Typo: s/However/However,/

Buristan commented on test/treegen.lua:
```diff
@@ -0,0 +1,129 @@
+end
+
+
+function treegen.before_all_default(g)
```
`before_all_default()`/`after_all_defalt()` is slightly misleading to me.
Maybe `before_all_init()`/`after_all_clean()` is better?
Feel free to ignore.

Totktonada replied:
Agreed to name it `init` and `clean`.

Totktonada replied:
Renamed.

Buristan reviewed (commented):
**[PATCH 4/7] test: add a generator for a tree of Lua files**

LGTM, except a few insignificant nits below.

> based test. Also it gain ability to provide several templates for Lua

Typo: s/Also it gain/Also, it gains/

Buristan reviewed (commented):
**[PATCH 5/7] test: add a simple tarantool runner/output catcher**
LGTM, except a single nit regarding the commit message.

> However there is nothing like 'just run tarantool with given args and

Typo: s/However/However,/

Totktonada commented on src/lua/init.c:
```diff
@@ -839,6 +839,13 @@ tarantool_lua_init(const char *tarantool_bin, int argc, char **argv)
+	 * Similar to _LOADED (package.loaded).
+	 */
+	lua_createtable(L, 0, 128);
+	lua_setfield(L, LUA_REGISTRYINDEX, "_TARANTOOL_BUILTIN");
```
I think it would be a bit cumbersome to include a header just to use the name. If you don't mind, I'll pass it over.

Buristan commented on test/luajit-test-init.lua:
```diff
@@ -100,6 +100,45 @@ rawset(_G, 'pairs', pairs_M.builtin_pairs)
+-- The test case expects that `foo` contains a return value of
+-- libs/foo.lua.
+--
+-- However when the override loader is enabled, the following
```
Typo: s/However/However,/

Totktonada replied:
Fixed. Resolved the discussion as trivial.

Buristan commented on test/luajit-test-init.lua:
```diff
@@ -100,6 +100,45 @@ rawset(_G, 'pairs', pairs_M.builtin_pairs)
+-- Unlikely there is a valid user scenario for a package.path
+-- entry without the interrogation mark. We consider this override
+-- loader behavior as valid. The loader is disabled to pass the
+-- test suite, but it is likely would be more correct to discard
```
Typo? s/it is likely would be/likely it would be/

Totktonada replied:
Seems so. Fixed. Resolved the discussion as trivial.

Buristan reviewed (commented):
**[PATCH 6/7] lua: allow to override a built-in module**
Generally, LGTM, just minor nits and questions below.

Buristan reviewed (approved):
**[PATCH 7/7] lua: enable/disable override loader**
First of all want to say that I really love `loaders` internal functions. They look very nice (`conditional_loader`, `chain_loader` are state of art for me) :).

Ignoring value of `TT_OVERRIDE_BUILTIN` when it can't be parsed looks fragile, but OK for me as far as it is more like for internal usage.

Thanks for the patch-set! Nice work!
LGTM!

Totktonada commented on test/app-luatest/override_test.lua:
```diff
@@ -0,0 +1,146 @@
+-- - misc -- tarantool's module implemented as part of LuaJIT.
+-- - *.lib, internal.*, utils.* and so on -- tarantool internal
+--   modules
+-- - memprof.*, misc.*, sysprof.*, table.*, timezones -- unclear
```
Seems so.

```lua
tarantool> do local modules = {} for m in pairs(package.loaded) do if m:startswith('misc.') then table.insert(modules, m) end end table.sort(modules) return modules end
---
- - misc.memprof
  - misc.sysprof
...
```

Totktonada commented on test/app-luatest/override_test.lua:
```diff
@@ -0,0 +1,146 @@
+    'iconv',
+    'key_def',
+    'luadebug',
+    'memprof',
```
It works, so it is ok to have there. If you're about `memprof.*` exclusion, it is glob, not regex :)

Totktonada commented on src/lua/init.c:
```diff
@@ -839,6 +839,13 @@ tarantool_lua_init(const char *tarantool_bin, int argc, char **argv)
+	 * Create a table for storing loaded built-in modules.
+	 * Similar to _LOADED (package.loaded).
+	 */
+	lua_createtable(L, 0, 128);
```
Changed to `lua_newtable(L)`.

Totktonada commented on test/unit/luaT_tuple_new.c:
```diff
@@ -187,6 +187,10 @@ main()
 	struct lua_State *L = luaL_newstate();
 	luaL_openlibs(L);
 
+	/* luaT_newmodule() assumes that this table exists. */
```
I added a helper named `luaT_newteststate()`. I'm not sure that it is what you asked, but it makes the code more clean.

Totktonada commented on test/app-luatest/override_test.lua:
```diff
@@ -0,0 +1,146 @@
+-- A couple of test cases with overriding built-in modules.
+--
+-- In a general case it is not a trivial task to correctly
+-- override a built-in module. However there are modules that
```
Fixed. Resolved the discussion as trivial.

Buristan commented on test/app-luatest/override_test.lua:
```diff
@@ -0,0 +1,146 @@
+    'pwd',
+    'socket',
+    'swim',
+    'sysprof',
```
It’s OK, see [reply in this discussion](https://github.com/tarantool/tarantool/pull/8300#discussion_r1109540739).

igormunkin reviewed (approved):
**[PATCH 6/7] lua: allow to override a built-in module**

@Totktonada, thanks for the patch! LGTM.

igormunkin reviewed (approved):
**[PATCH 7/7] lua: enable/disable override loader**

@Totktonada, thanks for the patch! LGTM.

Totally agree with @Buristan regarding `conditional_loader` :)

Totktonada commented:
box-luatest/gh_8192_feedback_daemon_metrics_test.lua fails on the coverage job in CI, but passes locally on a debug build. I'm looking into it.

Totktonada commented:
> box-luatest/gh_8192_feedback_daemon_metrics_test.lua fails on the coverage job in CI, but passes locally on a debug build. I'm looking into it.

I decided to adjust a test. See details below.

@drewdzzz, please take a look if time permits.

I have added one commit. Copy-pasted its description.

> ### test: expect slow metrics collection on ASAN build
>
> The feedback daemon test case fails constantly in CI after recent changes for override modules. I found no direct correspondence between the changes and the feedback daemon code/test case.
>
> The test case runs metrics collection for 0.3 seconds with 0.01 interval. It accepts [22; 38] samples as a correct result. I got 7-22 samples on ASAN build in CI.
>
> It means that the tested code is working, but slower. The tested code don't run the newly added code, but the new code is run at tarantool initialization. I think that the fails in CI are due to influence of the changed initialization code in composition with LuaJIT's unstable performance: some traces are collected differently, GC runs on different points and we're here.
>
> I performed a brief comparison of amount of fails in the test in parallel run (8 logical cores, 16 runs in parallel) of baseline (master) and changed code with and without ASAN. I can't achieve results that would be stable enough for judgements. At least, I can't do it without a long running testing.
>
> I changed the feedback daemon test case to accept [5; 55] samples as a correct result, when tarantool is built with ASAN. The old [22; 38] range is accepted for other builds, as before

<details>
<summary>Patch (collapsed)</summary>

```diff
diff --git a/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua b/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
index 1c278c7b5..c5508040c 100644
--- a/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
+++ b/test/box-luatest/gh_8192_feedback_daemon_metrics_test.lua
@@ -27,6 +27,7 @@ g.test_metrics = function()
     g.server:exec(function()
         local t = require('luatest')
         local fiber = require('fiber')
+        local tarantool = require('tarantool')
         local m = {}
         -- Method checks that we pass required arguments to collect defaults.
         -- Note that if one of assertions in the method fails, you will not
@@ -45,8 +46,12 @@ g.test_metrics = function()
         local fb = daemon.generate_feedback()
         -- This check may fail because of assertion failure in collect method.
         t.assert_type(fb.metrics, 'table')
-        -- Huge margin to prevent flakiness on slow environments like coverity.
-        t.assert_almost_equals(#fb.metrics, 30, 8)
+        -- Huge margin to prevent flakiness.
+        --
+        -- Even bigger for the ASAN build, which is known as slow: just several
+        -- samples are okay for this build type.
+        local margin = tarantool.build.flags:find('-fsanitize') and 25 or 8
+        t.assert_almost_equals(#fb.metrics, 30, margin)
         fb = daemon.generate_feedback()
         -- Check is metrics were reset. Fiber yields on feedback generation, so
         -- metrics may be not empty.
```
</details>

Totktonada commented:
`checks` integration test needs https://github.com/tarantool/checks/pull/51.

Totktonada commented:
Update regarding the failed test: I refused the idea to conditionally extend the range of allowed samples count. The test fails on macOS in CI too and it seems that the only way to proceed with the patchset now is to accept nearly any amount of metrics samples in the test on any build type.
